README.md
# minutes-to-max-human-time
[![Build Status via Travis CI](https://travis-ci.org/tonybranfort/minutes-to-max-human-time.svg?branch=master)](https://travis-ci.org/tonybranfort/req-uscis-status)
[![Coverage Status](https://coveralls.io/repos/github/tonybranfort/minutes-to-max-human-time/badge.svg?branch=master)](https://coveralls.io/github/tonybranfort/minutes-to-max-human-time?branch=master)

Takes minutes and returns the maximum whole human readable equivalent in minutes, hours, days or weeks.  Mimics Google Calendar online appointments for setting reminders (notifications).  

For example: 
* 50 minutes returns 50 minutes
* 60 minutes returns 1 hour
* 70 minutes returns 70 minutes
* 1440 minutes returns 1 day
* 1441 minutes returns 1441 minutes
* 100800 minutes returns 1 week  

Install options: 
* `npm install minutes-to-max-human-time`
* `bower install minutes-to-max-human-time`

###Example: 

```
var m2MaxHT = require('minutes-to-max-human-time');

m2MaxHT.getMax(0);    // { timeQty: 0, timePeriod: 'minutes' }

m2MaxHT.getMax(59);   // { timeQty: 59, timePeriod: 'minutes' }

m2MaxHT.getMax(60);   // { timeQty: 1, timePeriod: 'hours' }

m2MaxHT.getMax(61);   // { timeQty: 61, timePeriod: 'minutes' }

m2MaxHT.getMax(60*23);  // { timeQty: 23, timePeriod: 'hours' }

m2MaxHT.getMax(60*24);  // { timeQty: 1, timePeriod: 'days' }

m2MaxHT.getMax(60*25);  // { timeQty: 25, timePeriod: 'hours' }

m2MaxHT.getMax(60*24*6); // { timeQty: 6, timePeriod: 'days' }

m2MaxHT.getMax(60*24*7+1);
// { timeQty: 10081, timePeriod: 'minutes' }

m2MaxHT.getMax(60*24*7);  // { timeQty: 1, timePeriod: 'weeks' }

m2MaxHT.getMax(60*24*8);  // { timeQty: 8, timePeriod: 'days' }
```

minutes-to-max-human-time has 2 functions: `getMax` and `getMinutes`. 

###getMax(minutes, objectToUpdate, maxMinutes)
Returns an object with these 2 parameters: 
*  timeQty: integer,
*  timePeriod: string of 'minutes', 'hours', 'days',  or 'weeks'

'weeks' is the maximum whole human readable equivalent given minutes.  

####`minutes`
* integer, required

####`objectToUpdate`
* object literal, optional.  
* If passed in `objectToUpdate` will be updated with timeQty and timePeriod parameters.  If not included, a new object is returned. 

example: 
```
var m2MaxHT = require('minutes-to-max-human-time');
var myObj = {method: 'email', minutes:60};
m2MaxHT.getMax(myObj.minutes, myObj);
console.log(myObj);
// { method: 'email', minutes: 60, timeQty: 1, timePeriod: 'hours' }

```

####`maxMinutes`
* integer, optional, default = 40320 (4 weeks);  
* If passed in will reset `minutes` parameter to `maxMinutes` if `minutes` is > `maxMinutes`

example: 
```
m2MaxHT.getMax(85, {}, 80);
// { timeQty: 80, timePeriod: 'minutes' }  // minutes was set to 80 from 85

```

###getMinutes(object)
Returns the number of minutes given an object with parameters `timeQty` and `timePeriod`. 
####`object`
* object literal, required. 
* must be an object with parameters `timeQty` and `timePeriod` where timeQty is an integer and timePeriod is a string of 'minutes','hours','days' or 'weeks'. 

example: 
```
var m2MaxHT = require('minutes-to-max-human-time');

var myObj = {timeQty: 1, timePeriod: 'hours' };
var minutes = m2MaxHT.getMinutes(myObj);
console.log(minutes);
// 60

```




CHANGELOG.md
# v1.0.0
- Initial release with 2 functions `getMax`, `getMinutes`.


test/test.js
var gcrmt = require('../lib/index.js');
var expect = require('expect.js'); 

describe('Google calendar reminder max time ',function(){
  describe('getMax', function() {
    it('should return minutes when no match to hours, days, weeks',function(){
      expect(gcrmt.getMax(0).timeQty).to.be(0); 
      expect(gcrmt.getMax(0).timePeriod).to.be('minutes'); 
      expect(gcrmt.getMax(1).timeQty).to.be(1); 
      expect(gcrmt.getMax(1).timePeriod).to.be('minutes'); 
      expect(gcrmt.getMax(59).timeQty).to.be(59); 
      expect(gcrmt.getMax(59).timePeriod).to.be('minutes'); 

      expect(gcrmt.getMax(60).timePeriod).to.not.be('minutes'); 

      expect(gcrmt.getMax(61).timeQty).to.be(61); 
      expect(gcrmt.getMax(61).timePeriod).to.be('minutes'); 

      expect(gcrmt.getMax(61).timeQty).to.be(61); 
      expect(gcrmt.getMax(61).timePeriod).to.be('minutes'); 

      // test +/ 2 hours
      expect(gcrmt.getMax(60*2-1).timeQty).to.be(60*2-1); 
      expect(gcrmt.getMax(60*2-1).timePeriod).to.be('minutes'); 

      expect(gcrmt.getMax(60*2).timePeriod).to.not.be('minutes'); 

      expect(gcrmt.getMax(60*2+1).timeQty).to.be(60*2+1); 
      expect(gcrmt.getMax(60*2+1).timePeriod).to.be('minutes'); 

      // test +/ 1 day
      expect(gcrmt.getMax(60*24-1).timeQty).to.be(60*24-1); 
      expect(gcrmt.getMax(60*24-1).timePeriod).to.be('minutes'); 

      expect(gcrmt.getMax(60*24).timePeriod).to.not.be('minutes'); 

      expect(gcrmt.getMax(60*24+1).timeQty).to.be(60*24+1); 
      expect(gcrmt.getMax(60*24+1).timePeriod).to.be('minutes'); 

      // test +/ 1 week
      expect(gcrmt.getMax(60*24*7-1).timeQty).to.be(60*24*7-1); 
      expect(gcrmt.getMax(60*24*7-1).timePeriod).to.be('minutes'); 

      expect(gcrmt.getMax(60*24*7).timePeriod).to.not.be('minutes'); 

      expect(gcrmt.getMax(60*24*7+1).timeQty).to.be(60*24*7+1); 
      expect(gcrmt.getMax(60*24*7+1).timePeriod).to.be('minutes'); 

    }); 

    it('should return hours when no match to days, weeks',function(){
      expect(gcrmt.getMax(60).timeQty).to.be(1); 
      expect(gcrmt.getMax(60).timePeriod).to.be('hours'); 
      expect(gcrmt.getMax(60*2).timeQty).to.be(2); 
      expect(gcrmt.getMax(60*2).timePeriod).to.be('hours'); 

      // test +/- 1 day
      expect(gcrmt.getMax(60*23).timeQty).to.be(23); 
      expect(gcrmt.getMax(60*23).timePeriod).to.be('hours'); 

      expect(gcrmt.getMax(60*24).timePeriod).to.not.be('hours'); 

      expect(gcrmt.getMax(60*25).timeQty).to.be(25); 
      expect(gcrmt.getMax(60*25).timePeriod).to.be('hours'); 

      // test +/- 2 days
      expect(gcrmt.getMax(60*47).timeQty).to.be(47); 
      expect(gcrmt.getMax(60*47).timePeriod).to.be('hours'); 

      expect(gcrmt.getMax(60*24*2).timePeriod).to.not.be('hours'); 

      expect(gcrmt.getMax(60*49).timeQty).to.be(49); 
      expect(gcrmt.getMax(60*49).timePeriod).to.be('hours'); 

      // test +/- 1 week
      expect(gcrmt.getMax(60*(24*7-1)).timeQty).to.be(24*7-1); 
      expect(gcrmt.getMax(60*(24*7-1)).timePeriod).to.be('hours'); 

      expect(gcrmt.getMax(60*24*7).timePeriod).to.not.be('hours'); 

      expect(gcrmt.getMax(60*(24*7+1)).timeQty).to.be(24*7+1); 
      expect(gcrmt.getMax(60*(24*7+1)).timePeriod).to.be('hours'); 

      // test +/- 4 weeks
      expect(gcrmt.getMax(60*(24*28-1)).timeQty).to.be(24*28-1); 
      expect(gcrmt.getMax(60*(24*28-1)).timePeriod).to.be('hours'); 

      expect(gcrmt.getMax(60*24*28).timePeriod).to.not.be('hours'); 


    }); 

    it('should return days when no match to weeks',function(){
      expect(gcrmt.getMax(60*24).timeQty).to.be(1); 
      expect(gcrmt.getMax(60*24).timePeriod).to.be('days'); 

      expect(gcrmt.getMax(60*24*2).timeQty).to.be(2); 
      expect(gcrmt.getMax(60*24*2).timePeriod).to.be('days'); 

      expect(gcrmt.getMax(60*24*6).timeQty).to.be(6); 
      expect(gcrmt.getMax(60*24*6).timePeriod).to.be('days'); 

      expect(gcrmt.getMax(60*24*7).timePeriod).to.not.be('days'); 

      expect(gcrmt.getMax(60*24*8).timeQty).to.be(8); 
      expect(gcrmt.getMax(60*24*8).timePeriod).to.be('days'); 

      // test +/- 4 weeks
      expect(gcrmt.getMax(60*24*(7*4-1)).timeQty).to.be(7*4-1); 
      expect(gcrmt.getMax(60*24*(7*4-1)).timePeriod).to.be('days'); 

      expect(gcrmt.getMax(60*24*(7*4)).timePeriod).to.not.be('days'); 


    }); 

    it('should return weeks when exact minutes match to weeks',function(){

      expect(gcrmt.getMax(60*24*7*1).timeQty).to.be(1); 
      expect(gcrmt.getMax(60*24*7*1).timePeriod).to.be('weeks'); 

      expect(gcrmt.getMax(60*24*7*2).timeQty).to.be(2); 
      expect(gcrmt.getMax(60*24*7*2).timePeriod).to.be('weeks'); 

      expect(gcrmt.getMax(60*24*7*3).timeQty).to.be(3); 
      expect(gcrmt.getMax(60*24*7*3).timePeriod).to.be('weeks'); 

      expect(gcrmt.getMax(60*24*7*4).timeQty).to.be(4); 
      expect(gcrmt.getMax(60*24*7*4).timePeriod).to.be('weeks'); 

    }); 

    it('should update object if it is passed into getMax', function() {
      var obj = {'blah': 40}; 
      var to = gcrmt.getMax(3, obj);
      expect(obj.timeQty).to.be(3);  
      expect(obj.timePeriod).to.be('minutes');  
      expect(obj.blah).to.be(40); 
    }); 

    it('should allow 40320 minutes max by default', function() {
      expect(gcrmt.getMax(40321).timeQty).to.be(4);  
      expect(gcrmt.getMax(40321).timePeriod).to.be('weeks');  
    }); 

    it('should allow adjust max minutes with parameter', function() {
      expect(gcrmt.getMax(40321,{},40321).timeQty).to.be(40321);  
      expect(gcrmt.getMax(40321,{},40321).timePeriod).to.be('minutes');  

      expect(gcrmt.getMax(40319,{},40321).timeQty).to.be(40319);  
      expect(gcrmt.getMax(40319,{},40321).timePeriod).to.be('minutes');  

      expect(gcrmt.getMax(40321,{},59).timeQty).to.be(59);  
      expect(gcrmt.getMax(40321,{},59).timePeriod).to.be('minutes');  

      expect(gcrmt.getMax(40321,{},60).timeQty).to.be(1);  
      expect(gcrmt.getMax(40321,{},60).timePeriod).to.be('hours');  
    }); 

  });

  describe('getMinutes', function() {
    it('should return correct minutes for passing in minutes', function() {
      expect(gcrmt.getMinutes({timeQty: 59, timePeriod: 'minutes'})).to.be(59); 
      expect(gcrmt.getMinutes({timeQty: 60, timePeriod: 'minutes'})).to.be(60); 
      expect(gcrmt.getMinutes({timeQty: 61, timePeriod: 'minutes'})).to.be(61); 
      expect(gcrmt.getMinutes({timeQty: 1441, timePeriod: 'minutes'})).to.be(1441); 
    }); 
    it('should return correct mintues for passing in hours', function() {
      expect(gcrmt.getMinutes({timeQty: 1, timePeriod: 'hours'})).to.be(60); 
      expect(gcrmt.getMinutes({timeQty: 5, timePeriod: 'hours'})).to.be(60*5); 
      expect(gcrmt.getMinutes({timeQty: 24, timePeriod: 'hours'})).to.be(60*24); 
    }); 

    it('should return correct mintues for passing in days', function() {
      expect(gcrmt.getMinutes({timeQty: 1, timePeriod: 'days'})).to.be(60*24*1); 
      expect(gcrmt.getMinutes({timeQty: 5, timePeriod: 'days'})).to.be(60*24*5); 
      expect(gcrmt.getMinutes({timeQty: 7, timePeriod: 'days'})).to.be(60*24*7); 
    }); 

    it('should return correct mintues for passing in weeks', function() {
      expect(gcrmt.getMinutes()).not.to.be.NaN; 
      expect(gcrmt.getMinutes({timeQty: 1, timePeriod: 'weeks'})).to.be(60*24*7*1); 
      expect(gcrmt.getMinutes({timeQty: 2, timePeriod: 'weeks'})).to.be(60*24*7*2); 
      expect(gcrmt.getMinutes({timeQty: 4, timePeriod: 'weeks'})).to.be(60*24*7*4); 
    }); 

    it('should return NaN if invalid time object passed in', function() {
      // expect library uses === for to.be and to.equal assertion so 
      //   expect(NaN) won't work directly so using goofy convert to string
      expect(gcrmt.getMinutes() + '').to.be('NaN'); 
      expect(gcrmt.getMinutes({}) + '').to.be('NaN');
      expect(gcrmt.getMinutes({timeQty: 30}) + '').to.be('NaN');
      expect(gcrmt.getMinutes({timeQty: 30, timePeriod: 'hours'}) + '').not.to.be('NaN');
      expect(gcrmt.getMinutes({timeQty: 30, timePeriod: 'monkey'}) + '').to.be('NaN');
      expect(gcrmt.getMinutes({timeQty: 30, timePeriod: 'day'}) + '').to.be('NaN');
    }); 

  });  // end of getMinutes()
});




lib/index.js
// takes minutes qty(integer) and returns on object: 
//  {timeQty: int,
//   timePeriod: ('minutes', 'hours', 'days', 'weeks')}
// where timePeriod is the maximum possible without a partial of that timePeriod
// 50 minutes should display 50 minutes
// 60 minutes should display 1 hour
// 70 minutes should display 70 minutes
// 1440 minutes should display 1 day
// 1441 minutes should display 1440 minutes
// 6 days should display 6 days
// 7 days should display 1 week 
// 8 days should display 8 days
// etc 

// (function(root) {

(function () {

  var root = this; 

  var minutesToMaxHumanTime = {getMax: getMax, getMinutes: getMinutes}; 

  var MINUTES_IN_HOUR = 60; 
  var MINUTES_IN_DAY = 60*24; 
  var MINUTES_IN_WEEK = 60*24*7; 

  // thanks async: 
  // Export the minutesToMaxHumanTime object for **Node.js**, with
  // backwards-compatibility for the old `require()` API. If we're in
  // the browser, add `minutesToMaxHumanTime` as a global object.
  if (typeof exports !== 'undefined') {
    if (typeof module !== 'undefined' && module.exports) {
      exports = module.exports = minutesToMaxHumanTime;
    }
    exports.minutesToMaxHumanTime = minutesToMaxHumanTime;
  } else {
    root.minutesToMaxHumanTime = minutesToMaxHumanTime;
  }

  function getMax(minutes, returnObj, maxMinutes) {

    var timeQty; 

    returnObj = returnObj && isObjectLiteral(returnObj) ? returnObj : {}; 

    maxMinutes = maxMinutes ? maxMinutes : 40320; 

    minutes = minutes > maxMinutes ? maxMinutes : minutes; 

    // if minutes is 0, then should return 0  minutes
    if(minutes === 0) { 
      updateReturnObj(minutes, 'minutes'); 
      return returnObj; 
    }

    if(minutes % MINUTES_IN_WEEK === 0 ) {
      timeQty = minutes / MINUTES_IN_WEEK; 
      updateReturnObj(timeQty, 'weeks'); 
      return returnObj; 
    }

    if(minutes % MINUTES_IN_DAY === 0 ) {
      timeQty = minutes / MINUTES_IN_DAY; 
      updateReturnObj(timeQty, 'days'); 
      return returnObj; 
    }

    if(minutes % MINUTES_IN_HOUR === 0 ) {
      timeQty = minutes / MINUTES_IN_HOUR; 
      updateReturnObj(timeQty, 'hours'); 
      return returnObj; 
    }

    updateReturnObj(minutes, 'minutes'); 
    return returnObj; 

    function updateReturnObj(timeQty, timePeriod) {
      // updates return object with timeQty and timePeriod
      returnObj.timeQty = timeQty; 
      returnObj.timePeriod = timePeriod; 
    }

  }

  function isObjectLiteral(obj) {
    // returns true if obj is an object literal: {} 
    // returns false if anything else including arrays and functions

    return Object.prototype.toString.call(obj) === '[object Object]'; 
  }

  function getMinutes(timeObj) {
    // returns minutes given a timeObj that is fits defintion of 
    //   object returned from getMax function 
    // returns NaN if timeObj doesn't fit definition

    timeObj = timeObj ? timeObj : {}; 

    var minutesIn = {
      minutes: 1,
      hours: 60,
      days: 60 * 24,
      weeks: 60 * 24 * 7
    }; 

    return timeObj.timeQty * minutesIn[timeObj.timePeriod]; 
  }

// return minutesToMaxHumanTime; 

}).call(this);

